File: techDose/Graph/course2.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        vector<vector<int>> adj(numCourses);
        vector<int> visited(numCourses, 0);
        vector<int> order;

        // Build adjacency list
        for (auto& pre : prerequisites) {
            adj[pre[1]].push_back(pre[0]);
        }

        // Perform DFS for each course
        for (int i = 0; i < numCourses; ++i) {
            if (visited[i] == 0) {
                if (!dfs(i, adj, visited, order)) {
                    return {}; // Return empty if a cycle is detected
                }
            }
        }

        reverse(order.begin(), order.end()); // Reverse to get the correct order
        return order;
    }

private:
    bool dfs(int course, vector<vector<int>>& adj, vector<int>& visited, vector<int>& order) {
        if (visited[course] == 1) return false; // Cycle detected
        if (visited[course] == 2) return true;  // Already processed, skip

        visited[course] = 1; // Mark as in-progress

        for (int next : adj[course]) {
            if (!dfs(next, adj, visited, order)) {
                return false; // Cycle detected
            }
        }

        visited[course] = 2; // Mark as fully processed
        order.push_back(course); // Add course to order in post-order
        return true;
    }
};

int main() {
    Solution sol;
    vector<vector<int>> prerequisites = {{1, 0}, {2, 0}, {3, 1}, {3, 2}};
    vector<int> order = sol.findOrder(4, prerequisites);

    for (int course : order) {
        cout << course << " ";
    }

    return 0;
}

File: techDose/Graph/course2.js
class Solution {
  isACyclic(adj, visited, curr, res) {
    if (visited[curr] === 2) {
      return [];
    }
    if (visited[curr] == 1) return res;

    visited[curr] = 2;

    for (let next of adj[curr]) {
      if (this.isACyclic(adj, visited, next, res)?.length < 1) {
        return [];
      }
    }

    visited[curr] = 1;
    res.push(curr);
    return res?.length === adj?.length && res;
  }

  findOrder(numCourses, prerequisites) {
    const adj = Array.from({ length: numCourses }, () => []);
    const visited = Array(numCourses).fill(0);
    const res = [];

    for (let i = 0; i < prerequisites?.length; ++i) {
      adj[prerequisites[i][1]].push(prerequisites[i][0]);
    }

    for (let i = 0; i < numCourses; ++i) {
      if (visited[i] === 0) {
        const result = this.isACyclic(adj, visited, i, res);

        if (result.length === 0) {
          return [];
        }
      }
    }

    return res.reverse();
  }
}

// Testing the code
const sol = new Solution();
const prerequisites = [
  [1, 0],
  [2, 0],
  [3, 1],
  [3, 2],
];
const order = sol.findOrder(4, prerequisites);

console.log(order.join(" "));
